File: ReplayKitTutorial-ObjC/BroadcastUploadExtension/SampleHandler.m
//
//  SampleHandler.m
//  BroadcastUploadExtension
//
//  Created by KyuJin Kim on 2016. 12. 28..
//  Copyright © 2016년 KyuJin Kim. All rights reserved.
//


#import "SampleHandler.h"

//  To handle samples with a subclass of RPBroadcastSampleHandler set the following in the extension's Info.plist file:
//  - RPBroadcastProcessMode should be set to RPBroadcastProcessModeSampleBuffer
//  - NSExtensionPrincipalClass should be set to this class

@implementation SampleHandler

- (void)broadcastStartedWithSetupInfo:(NSDictionary<NSString *,NSObject *> *)setupInfo {
    // User has requested to start the broadcast. Setup info from the UI extension will be supplied.
}

- (void)broadcastPaused {
    // User has requested to pause the broadcast. Samples will stop being delivered.
}

- (void)broadcastResumed {
    // User has requested to resume the broadcast. Samples delivery will resume.
}

- (void)broadcastFinished {
    // User has requested to finish the broadcast.
}

- (void)processSampleBuffer:(CMSampleBufferRef)sampleBuffer withType:(RPSampleBufferType)sampleBufferType {
    
    switch (sampleBufferType) {
        case RPSampleBufferTypeVideo:
            // Handle audio sample buffer
            NSLog(@"Video");
            break;
        case RPSampleBufferTypeAudioApp:
            // Handle audio sample buffer for app audio
            NSLog(@"Audio");
            break;
        default:
            break;
    }
}

@end

File: ReplayKit-Swift/BroadcaseExtension/SampleHandler.swift
//
//  SampleHandler.swift
//  BroadcaseExtension
//
//  Created by KyuJin Kim on 2017. 2. 27..
//  Copyright © 2017년 KyuJin Kim. All rights reserved.
//

import ReplayKit
import VideoToolbox
import lf

//  To handle samples with a subclass of RPBroadcastSampleHandler set the following in the extension's Info.plist file:
//  - RPBroadcastProcessMode should be set to RPBroadcastProcessModeSampleBuffer
//  - NSExtensionPrincipalClass should be set to this class

class SampleHandler: RPBroadcastSampleHandler {
    private var broadcaster:RTMPBroadcaster = RTMPBroadcaster()

    override open func broadcastStarted(withSetupInfo setupInfo: [String : NSObject]?) {
        print("broadcastStarted")
        super.broadcastStarted(withSetupInfo: setupInfo)
        guard
            let endpointURL:String = setupInfo?["endpointURL"] as? String,
            let streamName:String = setupInfo?["streamName"] as? String else {
                return
        }
        broadcaster.streamName = streamName
        broadcaster.connect(endpointURL, arguments: nil)
    }
    
    override func broadcastPaused() {
        // User has requested to pause the broadcast. Samples will stop being delivered.
    }
    
    override func broadcastResumed() {
        // User has requested to resume the broadcast. Samples delivery will resume.
    }
    
    override func broadcastFinished() {
        // User has requested to finish the broadcast.
    }
    
    override func processSampleBuffer(_ sampleBuffer: CMSampleBuffer, with sampleBufferType: RPSampleBufferType) {
        switch sampleBufferType {
        case .video:
            if let description:CMVideoFormatDescription = CMSampleBufferGetFormatDescription(sampleBuffer) {
                let dimensions:CMVideoDimensions = CMVideoFormatDescriptionGetDimensions(description)
                broadcaster.stream.videoSettings = [
                    "width": dimensions.width,
                    "height": dimensions.height,
                    "profileLevel": kVTProfileLevel_H264_Baseline_AutoLevel,
                ]
            }
            broadcaster.appendSampleBuffer(sampleBuffer, withType: .video)
        case .audioApp:
            broadcaster.appendSampleBuffer(sampleBuffer, withType: .audio)
        case .audioMic:
            break
        }
    }
}
